src/libraries/Common/src/System/Net/Http/X509ResourceClient.cs

```diff
@@ -0,0 +1,307 @@
+                // Note: using a ConstructorInfo instead of Activator.CreateInstance, so the ILLinker can see the usage through the lambda method.
+                object? socketsHttpHandler = socketsHttpHandlerCtor.Invoke(null);
+                pooledConnectionIdleTimeoutProp.SetValue(socketsHttpHandler, TimeSpan.FromSeconds(PooledConnectionIdleTimeoutSeconds));
+                allowAutoRedirectProp.SetValue(socketsHttpHandler, false);
+                object? httpClient = httpClientCtor.Invoke(new object?[] { socketsHttpHandler });
+
+                return async (string uriString, CancellationToken cancellationToken, bool async) =>
+                {
+                    Uri uri = new Uri(uriString);
+
+                    if (!IsAllowedScheme(uri.Scheme))
+                    {
+                        return null;
+                    }
+
+                    // Equivalent of:
+                    // HttpRequestMessage requestMessage = new HttpRequestMessage() { RequestUri = new Uri(uri) };
+                    // HttpResponseMessage responseMessage = httpClient.Send(requestMessage, cancellationToken);
+                    // Note: using a ConstructorInfo instead of Activator.CreateInstance, so the ILLinker can see the usage through the lambda method.
+                    object requestMessage = httpRequestMessageCtor.Invoke(null);
+                    requestUriProp.SetValue(requestMessage, uri);
+                    object responseMessage;
+
+                    if (async)
+                    {
+                        Task sendTask = (Task)sendAsyncMethod.Invoke(httpClient, new object[] { requestMessage, cancellationToken })!;
+                        await sendTask.ConfigureAwait(false);
+                        responseMessage = taskOfHttpResponseMessageResultProp.GetValue(sendTask)!;
+                    }
+                    else
+                    {
+                        responseMessage = sendMethod.Invoke(httpClient, new object[] { requestMessage, cancellationToken })!;
+                    }
+
+                    int redirections = 0;
+                    Uri? redirectUri;
+                    bool hasRedirect;
+                    while (true)
+                    {
+                        int statusCode = (int)responseStatusCodeProp.GetValue(responseMessage)!;
+                        object responseHeaders = responseHeadersProp.GetValue(responseMessage)!;
+                        Uri? location = (Uri?)responseHeadersLocationProp.GetValue(responseHeaders);
+                        redirectUri = GetUriForRedirect((Uri)requestUriProp.GetValue(requestMessage)!, statusCode, location, out hasRedirect);
+                        if (redirectUri == null)
+                        {
+                            break;
+                        }
+
+                        ((IDisposable)responseMessage).Dispose();
+
+                        redirections++;
+                        if (redirections > MaxRedirections)
```

Comment:
I don't remember why we do all this custom redirect handling rather than using the built-in support.  Do you?

Reply:
Because we cannot permit HTTP -> HTTPS redirects. https://github.com/dotnet/runtime/issues/45010 has more background, and https://github.com/dotnet/runtime/pull/48221 has the first fix for it.

Reply:
We should really consider adding https://github.com/dotnet/runtime/issues/45364, @karelz

Reply:
> are you suggesting in .NET 7.0?

If we have time.